book_service/book_service.go

```diff
@@ -0,0 +1,113 @@
+package book_Service
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/daniyakubov/book_service_n/cache"
+	"github.com/daniyakubov/book_service_n/db_service"
+	"github.com/daniyakubov/book_service_n/models"
+	errors "github.com/fiverr/go_errors"
+)
+
+type BookService struct {
+	booksCache cache.Cache
+	dbHandler  db_service.DBHandler
+}
+
+func NewBookService(booksCache cache.Cache, dbHandler db_service.DBHandler) BookService {
+	return BookService{
+		booksCache: booksCache,
+		dbHandler:  dbHandler,
+	}
+}
+
+func (b *BookService) AddBook(ctx context.Context, book *models.Book, username string, routeName string) (string, error) {
+	body, err := json.Marshal(book)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("couldn't unmarshal result of book with id: %s, in AddBook function ", book.Id))
+	}
+
+	id, err := b.dbHandler.AddBook(ctx, body)
+	if err != nil {
+		return "", err
+	}
+
+	err = b.booksCache.AddAction(username, "Put", routeName)
```

Comment:
Let's try to think of a way to ״AddAction״  for each route without having to explicitly write it down

Reply:
I meant "from the outside"  kind of :)

Reply:
that was a tricky one, I used the gin.Engine "Use" function :)